apps/src/dance/ai/AiBlockPreview.tsx

```diff
@@ -16,40 +16,35 @@ const AiBlockPreview: React.FunctionComponent<AiBlockPreviewProps> = ({
 }) => {
   const blockPreviewContainerRef = useRef<HTMLSpanElement>(null);
   const workspaceRef = useRef<Workspace | null>(null);
+  const isRtl = useSelector((state: {isRtl: boolean}) => state.isRtl);
 
   // Create the workspace once the container has been rendered.
   useEffect(() => {
     if (!blockPreviewContainerRef.current) {
       return;
     }
 
-    const emptyBlockXml = Blockly.utils.xml.textToDom('<xml></xml>');
+    const xml = `
+      <xml>
+        <block type="Dancelab_setForegroundEffectExtended">
+          <field name="EFFECT">"${results.foregroundEffect}"</field>
+          <next>
+            <block type="Dancelab_setBackgroundEffectWithPaletteAI">
+              <field name="PALETTE">"${results.backgroundColor}"</field>
+              <field name="EFFECT">"${results.backgroundEffect}"</field>
+            </block>
+          </next>
+        </block>
+      </xml>
+      `;
+
+    const blocks = Blockly.utils.xml.textToDom(xml);
     workspaceRef.current = Blockly.BlockSpace.createReadOnlyBlockSpace(
       blockPreviewContainerRef.current,
-      emptyBlockXml,
-      {}
+      blocks,
+      {rtl: isRtl}
     );
-  }, [blockPreviewContainerRef]);
-
-  // Build out the blocks.
-  useEffect(() => {
-    if (!blockPreviewContainerRef.current || !workspaceRef.current) {
-      return;
-    }
-    const blocksSvg = generateAiEffectBlocksFromResult(
-      workspaceRef.current,
-      results
-    );
-    blocksSvg.forEach((blockSvg: BlockSvg) => {
-      blockSvg.initSvg();
-      blockSvg.render();
-    });
-    Blockly.svgResize(workspaceRef.current as WorkspaceSvg);
-
-    return () => {
-      workspaceRef.current?.clear();
```

Comment:
There's some cleanup that @sanchitmalhotra126 added here that I'm not totally sure if we need or not. Some discussion here:

https://github.com/code-dot-org/code-dot-org/pull/54615#discussion_r1378270293

Sanchit, do you have any recollection on the details here?

Reply:
Yeah, I'm not sure what the correct approach here would be - previously, we would create the workspace and add the blocks as two separate steps, so they were paired with two separate cleanup actions (`workspace.clear()` whenever the contents of the blocks change, and `workspace.dispose()` whenever the view was being unmounted). Now however it looks like we're both creating the workspace and setting up the blocks in one step. I think it might still be a good idea to clear the workspace as a cleanup action here (by returning a function from this useEffect that calls `workspaceRef.current.clear()`) - do we know what happens when this component re-renders/when `results` update?

Reply:
With this new code, we land in this code (and do `workspaceRef.current = Blockly.BlockSpace.createReadOnlyBlockSpace()`) each time we arrive at the results screen with some new blocks.  

Do we need to call `workspaceRef.current.clear()` if we are about to overwrite the value of `workspaceRef.current` anyway?  

Should we be calling `workspaceRef.current?.dispose()` before we overwrite a non-`null` `workspaceRef.current`?

I've tested with returning a function from this `useEffect` to call `workspaceRef.current.clear()` and does do the clear before creating the new read-only block space, so we can just do that if it's all we need.

Reply:
If re-renders don't spawn extra blocks, then I think we should be okay? And if it does work the same with `workspaceRef.current.clear()` as the cleanup function, I'd probably prefer to do that just to be safe. 

We don't want to cleanup with `dispose()` here because Blockly has logic to unmount the workspace's parent node from the DOM since it considers a parent-less workspace to be a top-level workspace. So we only want to call `dispose()` when we're unmounting this component (which is done in the last dependency-less useEffect hook)